lib/nfc/protocols/felica/felica_listener.c

```diff
@@ -49,6 +57,102 @@ const FelicaData* felica_listener_get_data(const FelicaListener* instance) {
+    memset(instance->requested_blocks, 0, sizeof(instance->requested_blocks));
+    const FelicaBlockListElement* item =
+        felica_listener_block_list_item_get_first(instance, request);
+    for(uint8_t i = 0; i < resp->block_count; i++) {
+        instance->requested_blocks[i] = item->block_number;
+        FelicaCommanReadBlockHandler handler =
+            felica_listener_get_read_block_handler(item->block_number);
+
+        handler(instance, item->block_number, i, resp);
+
+        item = felica_listener_block_list_item_get_next(instance, item);
+    }
+
+    bit_buffer_reset(instance->tx_buffer);
+    bit_buffer_append_bytes(instance->tx_buffer, (uint8_t*)resp, resp->header.length);
+    free(resp);
+
+    return felica_listener_frame_exchange(instance, instance->tx_buffer);
+}
+
+static FelicaError felica_listener_command_handler_write(
+    FelicaListener* instance,
+    const FelicaListenerGenericRequest* const generic_request) {
+    FURI_LOG_D(TAG, "Write cmd");
+
+    const FelicaListenerWriteRequest* request = (FelicaListenerWriteRequest*)generic_request;
+    const FelicaListenerWriteBlockData* data_ptr =
+        felica_listener_get_write_request_data_pointer(instance, generic_request);
+
+    FelicaListenerWriteCommandResponse* resp =
+        (FelicaListenerWriteCommandResponse*)malloc(sizeof(FelicaListenerWriteCommandResponse));
+    furi_check(resp);
+
+    resp->response_code = FELICA_LISTENER_RESPONSE_CODE_WRITE;
+    resp->idm = request->base.header.idm;
+    resp->length = sizeof(FelicaListenerWriteCommandResponse);
+
+    if(felica_listener_validate_write_request_and_set_sf(instance, request, data_ptr, resp)) {
+        const FelicaBlockListElement* item =
+            felica_listener_block_list_item_get_first(instance, request);
+        for(uint8_t i = 0; i < request->base.header.block_count; i++) {
+            FelicaCommandWriteBlockHandler handler =
+                felica_listener_get_write_block_handler(item->block_number);
+
+            handler(instance, item->block_number, &data_ptr->blocks[i]);
+
+            item = felica_listener_block_list_item_get_next(instance, item);
+        }
+        felica_wcnt_increment(instance->data);
+    }
+
+    bit_buffer_reset(instance->tx_buffer);
+    bit_buffer_append_bytes(instance->tx_buffer, (uint8_t*)resp, resp->length);
+    free(resp);
+
+    return felica_listener_frame_exchange(instance, instance->tx_buffer);
+}
+
+static FelicaError felica_listener_process_request(
+    FelicaListener* instance,
+    const FelicaListenerGenericRequest* generic_request) {
+    const uint8_t cmd_code = generic_request->header.code;
+    switch(cmd_code) {
+    case FELICA_CMD_READ_WITHOUT_ENCRYPTION:
+        return felica_listener_command_handler_read(instance, generic_request);
+    case FELICA_CMD_WRITE_WITHOUT_ENCRYPTION:
+        return felica_listener_command_handler_write(instance, generic_request);
+    default:
+        furi_crash("FeliCa incorrect command");
```

Comment:
May be we should print log with unsupported command here? I think it's not right to crash if we don't support a command yet

Reply:
Actually there are only 2 possible commands in Felica: Read and Write and no others, therefore I've added crash here in order to escalate situation when wrong command code was provided. In normal conditions such situation is not possible, unless user will put another command code directly (for example using proxmark felica raw command) and in such situation crash is more preferable, because it attracts user's attention.

Reply:
I see. However, if we want to emulate the card, we still shouldn't crash. Even if you send wrong command to original card, it will not brick. I think just adding error log is enough.